### PR TITLE
Copy input tensors

### PR DESCRIPTION
**Pull request by artemrd**
I noticed that when inputs are connected to CUDA operators input data is not copied which caused memory access errors in CUDA code.

### CODE REVIEW

File: onnxruntime/core/framework/transformer_memcpy.cc [hunk 1]
```diff
@@ -43,12 +43,21 @@ bool TransformerMemcpyImpl::ModifyGraph(const KernelRegistryManager& kernel_regi
   // for initializers shared by different providers, create dups
   ProcessInitializers();
 
+  for (auto arg : graph_.GetInputs())
+    BuildDefsMapping(arg, kernel_registries);
```
Comment (HectorSVC) [inline onnxruntime/core/framework/transformer_memcpy.cc]: Could you provide more information for the issue you encountered? Add a test case for example. This suppose not the right place for the fix.

Comment (artemrd) [reply onnxruntime/core/framework/transformer_memcpy.cc]: What is the right place to copy inputs to CUDA memory?

Comment (HectorSVC) [reply onnxruntime/core/framework/transformer_memcpy.cc]: please refer to [https://github.com/Microsoft/onnxruntime/blob/439dbbada9546fa41d8abfe09d94b7129f934fd4/onnxruntime/core/session/onnxruntime_c_api.cc#L416](https://github.com/Microsoft/onnxruntime/blob/439dbbada9546fa41d8abfe09d94b7129f934fd4/onnxruntime/core/session/onnxruntime_c_api.cc#L416)

---
In reply to: [251656562](https://github.com/Microsoft/onnxruntime/pull/395#discussion_r251656562) [](ancestors = 251656562)

Comment (HectorSVC) [reply onnxruntime/core/framework/transformer_memcpy.cc]: I still confused how/why it happened

---
In reply to: [251946583](https://github.com/Microsoft/onnxruntime/pull/395#discussion_r251946583) [](ancestors = 251946583,251656562)

File: onnxruntime/core/framework/transformer_memcpy.cc [hunk 2]
```diff
@@ -43,12 +43,21 @@ bool TransformerMemcpyImpl::ModifyGraph(const KernelRegistryManager& kernel_regi
...
     BuildDefsMapping(arg, kernel_registries);
 
   for (auto arg : non_provider_output_defs_)
     BuildDefsMapping(arg, kernel_registries);
 
+  for (auto arg : graph_.GetInputs())
+    if (provider_input_defs_.count(arg)) {
+      AddCopyNode(const_cast<onnxruntime::NodeArg*>(arg), true);
+      modified = true;
+    }
```
Comment (KeDengMS) [inline onnxruntime/core/framework/transformer_memcpy.cc]: Do we have a testcase?

Comment (artemrd) [reply onnxruntime/core/framework/transformer_memcpy.cc]: I have a model, but no unit test yet.

Comment (artemrd) [reply onnxruntime/core/framework/transformer_memcpy.cc]: Modified and re-enabled existing test.

File: onnxruntime/core/framework/transformer_memcpy.cc (hunk 1 above)
Comment (artemrd) [reply onnxruntime/core/framework/transformer_memcpy.cc]: As discussed this fix should be good for now.

File: onnxruntime/test/framework/cuda/fence_cuda_test.cc [hunk 3]
```diff
@@ -62,28 +62,30 @@ static common::Status LoadInferenceSessionFromModel(InferenceSession& session, o
 
 CREATE_INITIALIZER_FUNC(float, TensorProto_DataType_FLOAT, add_float_data)
 CREATE_INITIALIZER_FUNC(int64_t, TensorProto_DataType_INT64, add_int64_data)
-// TO DO: Figure out a way to enable it again
-TEST(CUDAFenceTests, DISABLED_PartOnCPU) {
```
Comment (KeDengMS) [inline onnxruntime/test/framework/cuda/fence_cuda_test.cc]: >DISABLED_PartOnCPU [](start = 21, length = 18)

This test was disabled before when graph is always serialized before session initialize, and node can have providers specified. I remember with serialization to ModelProto, session initialize won't see the execution provider and thus everything runs on CPU or GPU, no mix and match.

Have you checked if the code path you added is hit in this test?

Comment (artemrd) [reply onnxruntime/test/framework/cuda/fence_cuda_test.cc]: Yes I have checked.

**Comment by KeDengMS** [comment]
> static common::Status LoadInferenceSessionFromModel(InferenceSession& session, onnxruntime::Model& model) {

I don't understand. This function serializes the model to stream which strips off any provider assignment on nodes...

---
Refers to: onnxruntime/test/framework/cuda/fence_cuda_test.cc:43 in d560916. [](commit_id = d56091676ab8303196c2bc3b23663cbd8e61e8c2, deletion_comment = False)

**Comment by artemrd** [comment]
> > static common::Status LoadInferenceSessionFromModel(InferenceSession& session, onnxruntime::Model& model) {
> 
> I don't understand. This function serializes the model to stream which strips off any provider assignment on nodes...How could this test hit the input shared by CPU/GPU case?
> 
> Refers to: onnxruntime/test/framework/cuda/fence_cuda_test.cc:43 in [d560916](https://github.com/Microsoft/onnxruntime/commit/d56091676ab8303196c2bc3b23663cbd8e61e8c2). [](commit_id = [d560916](https://github.com/Microsoft/onnxruntime/commit/d56091676ab8303196c2bc3b23663cbd8e61e8c2), deletion_comment = False)

Hmm, Initialize() actually assigned MatMul to CPU and Add to CUDA so accidentally it still has desired behavior. Let me check.

**Comment by HectorSVC** [comment]
> We discussed this change offline. Currently the CopyInputsOutputs method doesn't support the case where one input is fed to 2 nodes on different devices. Hence the solution in this PR is ok for now although it's suboptimal because the copy to the GPU will happen every single time Run() is called. At some point and depending on the frequency of such use cases, this support should be added in the framework without incurring an additional cost at runtime.

I may need to do a perf test for the normal model which don't have this kind of scenario to see whether there are some significant performance impact.

**Pull request merged**